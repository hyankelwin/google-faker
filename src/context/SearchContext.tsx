import React, { createContext, useState, useCallback, useContext } from 'react';
import { Animal } from '../api/google-faker-api/search/interfaces/animal.interface';
import { getAnimals } from '../api/google-faker-api/search/google-faker.request';
import { getTimeOut } from '../utils/getTimeOutSimulateApi';
import { checkAnimals } from '../utils/getAnimalType';

interface ISearch {
  isLoading: boolean;
  error: boolean;
  search: string;
  animals: Animal[];
  animalSelected: Animal;
  showCard: boolean;
  visibilityModalOptions: boolean;
  visibilityModalSearch: boolean;
  inputIsEmpty: boolean;
  handleFindSeach(value: string): void;
  handleClearInputValue(): void;
  handleSetItemSelected(item: Animal): void;
  handleSetShowCard(): void;
  handleSetClearCard(): void;
  handleSetVisibiltyModalOptions(value: boolean): void;
  handleSetVisibiltyModalSearch(value: boolean): void;
  handleSetEmptyInputValue(value: boolean): void;
}

const SearchContext = createContext<ISearch>({} as ISearch);

const SearchProvider: React.FC = ({ children }) => {
  const [search, setSearch] = useState('');
  const [showCard, setShowCard] = useState(false);
  const [inputIsEmpty, setInputIsEmpty] = useState(true);
  const [visibilityModalOptions, setVisibilityModalOptions] = useState(false);
  const [visibilityModalSearch, setVisibilityModalSearch] = useState(false);
  const [isLoading, setLoading] = useState<boolean>(false);
  const [error, setError] = useState<boolean>(false);
  const [animals, setAnimals] = useState<Animal[]>([]);
  const [animalSelected, setAnimalSelected] = useState<Animal>({} as Animal);

  const handleSetItemSelected = useCallback((item: Animal) => {
    setAnimalSelected(item);
  }, []);

  const handleClearInputValue = useCallback(() => {
    setSearch('');
  }, []);

  const handleSetEmptyInputValue = useCallback(value => {
    setInputIsEmpty(value);
  }, []);

  const handleSetShowCard = useCallback(() => {
    setShowCard(true);
  }, []);

  const handleSetClearCard = useCallback(() => {
    setShowCard(false);
  }, []);

  const handleSetLocalStorage = useCallback((value: string) => {
    setShowCard(false);
    localStorage.setItem('search', value);
  }, []);

  const handleSetVisibiltyModalOptions = useCallback((value: boolean) => {
    setVisibilityModalOptions(value);
  }, []);

  const handleSetVisibiltyModalSearch = useCallback((value: boolean) => {
    setVisibilityModalSearch(value);
  }, []);

  const handleFindSeach = useCallback(
    async (value: string) => {
      setLoading(true);
      setError(false);
      setSearch(value);
      handleSetLocalStorage(value);

      await getTimeOut(10);

      if (!checkAnimals(value)) {
        setLoading(false);
        setError(true);
        return;
      }

      const response = await getAnimals(value);

      setAnimals(response);
      setLoading(false);
    },
    [handleSetLocalStorage],
  );

  return (
    <SearchContext.Provider
      value={{
        search,
        isLoading,
        error,
        animals,
        animalSelected,
        showCard,
        visibilityModalOptions,
        visibilityModalSearch,
        inputIsEmpty,
        handleFindSeach,
        handleClearInputValue,
        handleSetItemSelected,
        handleSetShowCard,
        handleSetClearCard,
        handleSetVisibiltyModalOptions,
        handleSetVisibiltyModalSearch,
        handleSetEmptyInputValue,
      }}
    >
      {children}
    </SearchContext.Provider>
  );
};

function useSearch(): ISearch {
  const context = useContext(SearchContext);

  return context;
}

export { SearchProvider, useSearch };
