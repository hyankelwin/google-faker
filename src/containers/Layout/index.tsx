import React from 'react';

import { useHistory } from 'react-router-dom';
import Header from '../../components/Header';

import { Container, Content } from './styles';
import Footer from '../../components/Footer';
import { useSearch } from '../../context/SearchContext';

interface LayoutProps {
  page: string;
  isError?: boolean;
}

const Layout: React.FC<LayoutProps> = ({ page, isError = false, children }) => {
  const { handleSetVisibiltyModalOptions } = useSearch();
  const history = useHistory();

  return (
    <Content page={page}>
      <Header
        onClickOptions={() => handleSetVisibiltyModalOptions(true)}
        onClickNavigate={() => history.goBack()}
        page={page}
      />
      <Container>{children}</Container>
      <Footer page={page} isError={isError} />
    </Content>
  );
};

export default Layout;
