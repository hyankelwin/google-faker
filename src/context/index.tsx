import React from 'react';

import { SearchProvider } from './SearchContext';

const AppProvider: React.FC = ({ children }) => (
  <SearchProvider>{children}</SearchProvider>
);

export default AppProvider;
