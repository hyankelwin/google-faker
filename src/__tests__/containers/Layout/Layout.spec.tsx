import { fireEvent, render, waitFor } from '@testing-library/react';
import { mount } from 'enzyme';
import React from 'react';
import { useSearch } from '../../../context/SearchContext';
import Layout from '../../../containers/Layout';

const mockedHistoryGoBack = jest.fn();

jest.mock('react-router-dom', () => {
  return {
    useHistory: () => ({
      goBack: mockedHistoryGoBack,
    }),
    Link: ({ children }: { children: React.ReactNode }) => children,
  };
});

jest.mock('../../../context/SearchContext');

describe('Layout container', () => {
  it('should be able to render Layout container with success', () => {
    (useSearch as unknown as jest.Mock).mockReturnValue({
      handleSetVisibiltyModalOptions: () => jest.fn(),
    });

    const props = {
      page: 'home',
    };

    const wrapper = mount(<Layout {...props} />);

    expect(wrapper.debug()).toMatchSnapshot();
  });

  it('should be able to simulate click on OptionsIcon on Layout container', async () => {
    (useSearch as unknown as jest.Mock).mockReturnValue({
      handleSetVisibiltyModalOptions: () => jest.fn(),
    });

    const props = {
      page: 'search',
    };

    const { getByTestId, asFragment } = render(<Layout {...props} />);

    const iconOptions = getByTestId('options-icon');

    fireEvent.click(iconOptions);

    await waitFor(() => {
      expect(asFragment()).toMatchSnapshot();
    });
  });

  it('should be able to simulate click on GoogleImage on Layout container', async () => {
    (useSearch as unknown as jest.Mock).mockReturnValue({
      handleSetVisibiltyModalOptions: () => jest.fn(),
    });

    const props = {
      page: 'search',
    };

    const { getByTestId, asFragment } = render(<Layout {...props} />);

    const googleImage = getByTestId('google-image');

    fireEvent.click(googleImage);

    await waitFor(() => {
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
