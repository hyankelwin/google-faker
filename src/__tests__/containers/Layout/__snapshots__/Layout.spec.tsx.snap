// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout container should be able to render Layout container with success 1`] = `
"<Layout page=\\"home\\">
  <styled.div page=\\"home\\">
    <div className=\\"sc-kpDqfm ljSbcm\\">
      <Header onClickOptions={[Function: onClickOptions]} onClickNavigate={[Function: onClickNavigate]} page=\\"home\\">
        <styled.div page=\\"home\\">
          <div className=\\"sc-aXZVg faNpxW\\">
            <div>
              <styled.p>
                <p className=\\"sc-dcJsrY cDJWQS\\">
                  Agile Content Frontend test
                </p>
              </styled.p>
            </div>
            <div>
              <styled.div page=\\"home\\">
                <div className=\\"sc-fqkvVR eJjnVN\\">
                  <IoApps size={20}>
                    <IconBase attr={{...}} size={20}>
                      <svg stroke=\\"currentColor\\" fill=\\"currentColor\\" strokeWidth=\\"0\\" viewBox=\\"0 0 512 512\\" className={[undefined]} style={{...}} height={20} width={20} xmlns=\\"http://www.w3.org/2000/svg\\">
                        <path d=\\"M104 160a56 56 0 1156-56 56.06 56.06 0 01-56 56zm152 0a56 56 0 1156-56 56.06 56.06 0 01-56 56zm152 0a56 56 0 1156-56 56.06 56.06 0 01-56 56zM104 312a56 56 0 1156-56 56.06 56.06 0 01-56 56zm152 0a56 56 0 1156-56 56.06 56.06 0 01-56 56zm152 0a56 56 0 1156-56 56.06 56.06 0 01-56 56zM104 464a56 56 0 1156-56 56.06 56.06 0 01-56 56zm152 0a56 56 0 1156-56 56.06 56.06 0 01-56 56zm152 0a56 56 0 1156-56 56.06 56.06 0 01-56 56z\\" />
                      </svg>
                    </IconBase>
                  </IoApps>
                  <styled.img src=\\"user.png\\" alt=\\"Image to simulate user logged\\">
                    <img src=\\"user.png\\" alt=\\"Image to simulate user logged\\" className=\\"sc-iGgWBj eckbzv\\" />
                  </styled.img>
                </div>
              </styled.div>
            </div>
          </div>
        </styled.div>
      </Header>
      <styled.div>
        <div className=\\"sc-dhKdcB bdpLCv\\" />
      </styled.div>
      <Footer page=\\"home\\" isError={false}>
        <styled.div page=\\"home\\" isError={false}>
          <div className=\\"sc-dAlyuH ezSzub\\">
            <p>
              ©Google 2021
            </p>
            <p>
              Version 
              0.1.0
            </p>
          </div>
        </styled.div>
      </Footer>
    </div>
  </styled.div>
</Layout>"
`;

exports[`Layout container should be able to simulate click on GoogleImage on Layout container 1`] = `
<DocumentFragment>
  <div
    class="sc-kpDqfm jKebxL"
  >
    <div
      class="sc-aXZVg ecOmaW"
    >
      <div>
        <div>
          <img
            alt="Image of google search"
            class="sc-gEvEer bhHMwT"
            data-testid="google-image"
            src="logo-google.png"
          />
          <form>
            <div
              class="sc-gsFSXq fvHuLm"
            >
              <div
                class="sc-imWYAI eNglSV"
              >
                <svg
                  fill="none"
                  height="20"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="20"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <circle
                    cx="11"
                    cy="11"
                    r="8"
                  />
                  <line
                    x1="21"
                    x2="16.65"
                    y1="21"
                    y2="16.65"
                  />
                </svg>
                <input
                  data-testid="input-search"
                  value=""
                />
              </div>
            </div>
          </form>
        </div>
      </div>
      <div>
        <div
          class="sc-eqUAAy eLIJBU"
        >
          <svg
            data-testid="options-icon"
            fill="currentColor"
            height="20"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 1024 1024"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M388.8 896.4v-27.198c.6-2.2 1.6-4.2 2-6.4 8.8-57.2 56.4-102.4 112.199-106.2 62.4-4.4 115.2 31.199 132.4 89.199 2.2 7.6 3.8 15.6 5.8 23.4v27.2c-.6 1.8-1.6 3.399-1.8 5.399-8.6 52.8-46.6 93-98.6 104.4-4 .8-8 2-12 3h-27.2c-1.8-.6-3.6-1.6-5.4-1.8-52-8.4-91.599-45.4-103.6-96.8-1.2-5-2.6-9.6-3.8-14.2zm252.4-768.797l-.001 27.202c-.6 2.2-1.6 4.2-1.8 6.4-9 57.6-56.8 102.6-113.2 106.2-62.2 4-114.8-32-131.8-90.2-2.2-7.401-3.8-15-5.6-22.401v-27.2c.6-1.8 1.6-3.4 2-5.2 9.6-52 39.8-86 90.2-102.2 6.6-2.2 13.6-3.4 20.4-5.2h27.2c1.8.6 3.6 1.6 5.4 1.8 52.2 8.6 91.6 45.4 103.6 96.8 1.201 4.8 2.401 9.4 3.601 13.999zm-.001 370.801v27.2c-.6 2.2-1.6 4.2-2 6.4-9 57.4-58.6 103.6-114.6 106-63 2.8-116.4-35.2-131.4-93.8-1.6-6.2-3-12.4-4.4-18.6v-27.2c.6-2.2 1.6-4.2 2-6.4 8.8-57.4 58.6-103.601 114.6-106.2 63-3 116.4 35.2 131.4 93.8 1.6 6.4 3 12.6 4.4 18.8z"
            />
          </svg>
        </div>
        <div
          class="sc-fqkvVR cMyPno"
        >
          <svg
            fill="currentColor"
            height="20"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M104 160a56 56 0 1156-56 56.06 56.06 0 01-56 56zm152 0a56 56 0 1156-56 56.06 56.06 0 01-56 56zm152 0a56 56 0 1156-56 56.06 56.06 0 01-56 56zM104 312a56 56 0 1156-56 56.06 56.06 0 01-56 56zm152 0a56 56 0 1156-56 56.06 56.06 0 01-56 56zm152 0a56 56 0 1156-56 56.06 56.06 0 01-56 56zM104 464a56 56 0 1156-56 56.06 56.06 0 01-56 56zm152 0a56 56 0 1156-56 56.06 56.06 0 01-56 56zm152 0a56 56 0 1156-56 56.06 56.06 0 01-56 56z"
            />
          </svg>
          <img
            alt="Image to simulate user logged"
            class="sc-iGgWBj eckbzv"
            src="user.png"
          />
        </div>
      </div>
    </div>
    <div
      class="sc-dhKdcB bdpLCv"
    />
    <div
      class="sc-dAlyuH bdRIeH"
    >
      <p>
        ©Google 2021
      </p>
      <p>
        Version 0.1.0
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Layout container should be able to simulate click on OptionsIcon on Layout container 1`] = `
<DocumentFragment>
  <div
    class="sc-kpDqfm jKebxL"
  >
    <div
      class="sc-aXZVg ecOmaW"
    >
      <div>
        <div>
          <img
            alt="Image of google search"
            class="sc-gEvEer bhHMwT"
            data-testid="google-image"
            src="logo-google.png"
          />
          <form>
            <div
              class="sc-gsFSXq fvHuLm"
            >
              <div
                class="sc-imWYAI eNglSV"
              >
                <svg
                  fill="none"
                  height="20"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="20"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <circle
                    cx="11"
                    cy="11"
                    r="8"
                  />
                  <line
                    x1="21"
                    x2="16.65"
                    y1="21"
                    y2="16.65"
                  />
                </svg>
                <input
                  data-testid="input-search"
                  value=""
                />
              </div>
            </div>
          </form>
        </div>
      </div>
      <div>
        <div
          class="sc-eqUAAy eLIJBU"
        >
          <svg
            data-testid="options-icon"
            fill="currentColor"
            height="20"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 1024 1024"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M388.8 896.4v-27.198c.6-2.2 1.6-4.2 2-6.4 8.8-57.2 56.4-102.4 112.199-106.2 62.4-4.4 115.2 31.199 132.4 89.199 2.2 7.6 3.8 15.6 5.8 23.4v27.2c-.6 1.8-1.6 3.399-1.8 5.399-8.6 52.8-46.6 93-98.6 104.4-4 .8-8 2-12 3h-27.2c-1.8-.6-3.6-1.6-5.4-1.8-52-8.4-91.599-45.4-103.6-96.8-1.2-5-2.6-9.6-3.8-14.2zm252.4-768.797l-.001 27.202c-.6 2.2-1.6 4.2-1.8 6.4-9 57.6-56.8 102.6-113.2 106.2-62.2 4-114.8-32-131.8-90.2-2.2-7.401-3.8-15-5.6-22.401v-27.2c.6-1.8 1.6-3.4 2-5.2 9.6-52 39.8-86 90.2-102.2 6.6-2.2 13.6-3.4 20.4-5.2h27.2c1.8.6 3.6 1.6 5.4 1.8 52.2 8.6 91.6 45.4 103.6 96.8 1.201 4.8 2.401 9.4 3.601 13.999zm-.001 370.801v27.2c-.6 2.2-1.6 4.2-2 6.4-9 57.4-58.6 103.6-114.6 106-63 2.8-116.4-35.2-131.4-93.8-1.6-6.2-3-12.4-4.4-18.6v-27.2c.6-2.2 1.6-4.2 2-6.4 8.8-57.4 58.6-103.601 114.6-106.2 63-3 116.4 35.2 131.4 93.8 1.6 6.4 3 12.6 4.4 18.8z"
            />
          </svg>
        </div>
        <div
          class="sc-fqkvVR cMyPno"
        >
          <svg
            fill="currentColor"
            height="20"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M104 160a56 56 0 1156-56 56.06 56.06 0 01-56 56zm152 0a56 56 0 1156-56 56.06 56.06 0 01-56 56zm152 0a56 56 0 1156-56 56.06 56.06 0 01-56 56zM104 312a56 56 0 1156-56 56.06 56.06 0 01-56 56zm152 0a56 56 0 1156-56 56.06 56.06 0 01-56 56zm152 0a56 56 0 1156-56 56.06 56.06 0 01-56 56zM104 464a56 56 0 1156-56 56.06 56.06 0 01-56 56zm152 0a56 56 0 1156-56 56.06 56.06 0 01-56 56zm152 0a56 56 0 1156-56 56.06 56.06 0 01-56 56z"
            />
          </svg>
          <img
            alt="Image to simulate user logged"
            class="sc-iGgWBj eckbzv"
            src="user.png"
          />
        </div>
      </div>
    </div>
    <div
      class="sc-dhKdcB bdpLCv"
    />
    <div
      class="sc-dAlyuH bdRIeH"
    >
      <p>
        ©Google 2021
      </p>
      <p>
        Version 0.1.0
      </p>
    </div>
  </div>
</DocumentFragment>
`;
